test/common/dynamo/dynamo_filter_test.cc

```diff
@@ -59,11 +60,21 @@ TEST_F(DynamoFilterTest, operatorPresent) {
   EXPECT_CALL(stats_, counter("prefix.dynamodb.operation.Get.upstream_rq_total_200"));
   EXPECT_CALL(stats_, counter("prefix.dynamodb.operation.Get.upstream_rq_total"));
 
-  EXPECT_CALL(stats_,
-              deliverTimingToSinks("prefix.dynamodb.operation.Get.upstream_rq_time_2xx", _));
-  EXPECT_CALL(stats_,
-              deliverTimingToSinks("prefix.dynamodb.operation.Get.upstream_rq_time_200", _));
-  EXPECT_CALL(stats_, deliverTimingToSinks("prefix.dynamodb.operation.Get.upstream_rq_time", _));
+  EXPECT_CALL(stats_, timer("prefix.dynamodb.operation.Get.upstream_rq_time_2xx"));
+  EXPECT_CALL(stats_, timer("prefix.dynamodb.operation.Get.upstream_rq_time_200"));
+  EXPECT_CALL(stats_, timer("prefix.dynamodb.operation.Get.upstream_rq_time"));
+  EXPECT_CALL(
+      stats_,
+      deliverTimingToSinks(
+          Property(&Stats::Metric::name, "prefix.dynamodb.operation.Get.upstream_rq_time_2xx"), _));
+  EXPECT_CALL(
+      stats_,
+      deliverTimingToSinks(
```

Comment:
It might help make the review easier if you can split out the timer/histogram changes from the tag extraction. LMK if that sounds (un)reasonable (I can share some `git` tips for doing surgery like this quickly).

Reply:
+1 to splitting that part out. Also, related to my other question as to whether we want to both with a full class hierarchy for histograms and timers.

Reply:
I think it makes sense to split out the changes not directly related to tags (Metric, class hierarchy,  and removing support for flushing histograms and timers without an explicit object) as the "first stage".  Once that gets merged, we can review the regex and tags portion that relies on the first stage.  Does that plan seem reasonable to y'all?

And in reference to your comment @mattklein123, I agree that we should condense them into a single class in the interest of code bloat.  Still thinking about how to do this.  Conceptually it seems like histogram is really the more general of the two types.  Either way, it makes sense that a simple flag or enum could differentiate between them.

Reply:
Yeah I was just thinking about this a bit and it gets ugly across a few different dimensions:
- should histograms/timers be in different stat namespaces (What if you have "foo" timer and "foo" histogram).
- the existence of the timespan class and allocateSpan() method

It might be better to just treat them as different things and do what you did. :(

The only other thing I can think of would be to:
- Rename Timer -> Histogram
- Remove Timespan class / allocateSpan() (this could be made into a helper class to snap a time and deliver it to a histogram when complete() is called).
- Have some method on a Histogram which tells what the type is (time, raw value, etc.)

I kind of like ^ but it would be a bunch of work.

Reply:
Yeah, I was running into a little bit of both of those.  I was going in the direction of your suggestion.  I think it's the cleanest implementation.  It might not be too bad.  Timespan actually wouldn't have to change much (or at all).  However, users would just have to directly construct a TimespanImpl, which adds the dependence on stats_impl for them instead of just using the stats interface classes.  Not sure if that matters too much, though.

As for the namespace issue, I think that's messier.  On one hand, we won't be sparing ourselves as much code if we still have to pipe them around separately in all of the stats management classes because they're considered to be distinct.  On the other hand, if they're in the same namespace, we would probably still make separate macros for them as to set the interpretation flag correctly.  This could definitely cause confusion in that that they would be the only two macros that can collide, and a collision would, AFAIU, happen silently and create a totally nonsensical combined output.

I'll give it a little more thought.

Reply:
> However, users would just have to directly construct a TimespanImpl, which adds the dependence on stats_impl for them instead of just using the stats interface classes. Not sure if that matters too much, though.

I think you could stick this in its own file, or even put it at the interface level since it's so tiny and just depends on the interface.

re: namespacing, IMO it's fine to have them in the same namespace. I agree it's confusing. I think you could likely treat a mismatch as a fatal error as its a programming error IMO that should be caught during tests.

Reply:
Sounds good.